external/webserver/http_client.c

```diff
@@ -210,6 +218,54 @@ static int process_trailer_header(char *buf, unsigned int buf_len, struct http_r
 	return 1;
 }
 
+static void parse_keep_alive_header(struct http_client_t *client, struct http_keyvalue_list_t *params)
+{
+	char *token = NULL;
+	char str_keep_alive[HTTP_CONF_MAX_VALUE_LENGTH] = { 0, };
+	char *header_field_value = NULL;
+
+	if (client->keep_alive_header_flag == 1) {
+		HTTP_LOGE("keep alive header is already set: timeout[%d] max[%d] \n",
+				client->keep_alive_timeout, client->max_request);
+		return;
+	}
+
+	header_field_value = http_keyvalue_list_find(params, "Keep-Alive");
+	if (strncmp(header_field_value, "(null)", 7) == 0) {
+		return;
+	}
+
+	HTTP_LOGD("Keep Alive Header [%s] \n", header_field_value);
+
+	strncpy(str_keep_alive, header_field_value, HTTP_CONF_MAX_VALUE_LENGTH);
+
+	token = strtok(str_keep_alive, ", ");
+	while (token != NULL) {
+		if(strncmp(token, "timeout=", 8) == 0) {
+			client->keep_alive_timeout = atoi(token + 8);
+			if (client->keep_alive_timeout < CONFIG_NETUTILS_WEBSERVER_MIN_CLIENT_RCV_TIMEOUT) {
+				client->keep_alive_timeout = CONFIG_NETUTILS_WEBSERVER_MIN_CLIENT_RCV_TIMEOUT;
+			} else if (client->keep_alive_timeout > CONFIG_NETUTILS_WEBSERVER_MAX_CLIENT_RCV_TIMEOUT) {
+				client->keep_alive_timeout = CONFIG_NETUTILS_WEBSERVER_MAX_CLIENT_RCV_TIMEOUT;
+			}
+
+			HTTP_LOGD("keep_alive_timeout : %d \n", client->keep_alive_timeout);
+		}
+
+		if(strncmp(token, "max=", 4) == 0) {
+			client->max_request = atoi(token + 4);
+			if (client->max_request < MIN_CLIENT_REQUEST) {
+				client->max_request =  MIN_CLIENT_REQUEST;
+			}
+
+			HTTP_LOGD("max request : %d \n", client->max_request);
+		}
+
+		token = strtok(NULL, ", ");
```

Comment:
Almost the same content as previous.
Even if first or second check is matched, there is no pointer moving. This causes redundant string operation.
Let's move the pointer when it's matched.
```
if (strncmp(token, "timeout=", 8) == 0) {
   ... = strtol((token + 8), &token, 10);
} else if (strncmp(token, "max=", ...) {
   ... = strtol((token + 4), &token, 10);
}
```

Reply:
done

Reply:
done

Reply:
What I suggest is using strtol directly instead of atoi. Because the `atoi` uses strtol internally, there is no more loop and no more complexity.
Because you use atoi, even `max=` is matched, pointer stays start of the argu. So that line 266 search `,` from start of the argu again. But if you use `strtol` directly with endptr, line 266 will not check the same string at all.

Reply:
When you find `max=` or `timeout=`, you can move the pointer to end of the argu. But current code doesn't so that it causes duplicate string check again.

Reply:
char * str[] = "timeout=500, max=1000";
        char *token = strtok(str, ", ");

        while (token != NULL)
        {
                if(strncmp(token, "timeout=", 8) == 0)   ==> at this point token is "timeout=500"
                {       
                        timeout = strtol((token + 8), &token, 10);   ==> at this point new token is ""

                } else if(strncmp(token, "max=", 4) == 0)
                {       
                        max = strtol((token + 4), &token, 10);
                }                                                                                                                                                             
          }
Then how loop will work ?

Reply:
I don't think so. Please check it with logs.

Reply:
#include <iostream>
#include <string.h>

using namespace std;


int main()
{
        char str[] = "timeout=500, max=1000";
        char *token = strtok(str, ", ");
        int timeout = 0;
        int max = 0;

        cout<<"token1:["<<token<<"]\n";                                                                                                                                                                     

        while (token != NULL)
        {
                cout<<"token2:["<<token<<"]\n";
                if(strncmp(token, "timeout=", 8) == 0)
                {       
                        timeout = strtol((token + 8), &token, 10); 
                        cout<<"token3:["<<token<<"]\n";

                } else if(strncmp(token, "max=", 4) == 0)
                {       
                        max = strtol((token + 4), &token, 10);
                        cout<<"token4:["<<token<<"]\n";
                }    
               token = strtok(NULL, ", "); 
        }

        return 0;
}
=================
Output
==============
$ g++ strtol_test.cpp -o st
$ ./st
token1:[timeout=500]
token2:[timeout=500]
token3:[]
token2:[max=1000]
token4:[]

Reply:
Oh you're right. I mis-understood it was strchr (finding next token). Please ignore my comment.